src/elm/Community.elm

```diff
@@ -405,12 +406,29 @@ encodeCreateCommunityData c =
         , ( "description", Encode.string c.description )
         , ( "inviter_reward", Eos.encodeAsset c.inviterReward )
         , ( "invited_reward", Eos.encodeAsset c.invitedReward )
+        , ( "min_balance", Eos.encodeAsset c.minBalance )
         , ( "has_objectives", Eos.encodeEosBool c.hasObjectives )
         , ( "has_shop", Eos.encodeEosBool c.hasShop )
         , ( "has_kyc", Eos.encodeEosBool c.hasKyc )
         ]
 
 
+createCommunityDataDecoder : Decoder CreateCommunityData
+createCommunityDataDecoder =
+    Decode.succeed CreateCommunityData
+        |> required "cmm_asset" Eos.decodeAsset
+        |> required "creator" Eos.nameDecoder
+        |> required "logo" Decode.string
+        |> required "name" Decode.string
+        |> required "description" Decode.string
+        |> required "inviter_reward" Eos.decodeAsset
+        |> required "invited_reward" Eos.decodeAsset
+        |> required "min_balance" Eos.decodeAsset
+        |> required "has_objectives" Eos.eosBoolDecoder
+        |> required "has_shop" Eos.eosBoolDecoder
+        |> required "has_kyc" Eos.eosBoolDecoder
+
+
```

Comment:
Why do we need a decoder for this?

Reply:
To create a new community, we use the `newCommunity` GraphQL subscription. So, if you look at [Community.Editor.Msg](https://github.com/cambiatus/frontend/pull/494/files#diff-4817c965f006ef54346aab990d3c1154f0573d7fb700e2ad7661c9c9e317ff84R574-R577), we fire a `Msg` when the user has `SubmittedForm`, which parses the model into a `CreateCommunityData` and, if it's valid, subscribes to the GraphQL `newCommunity` subscription, which then fires the `StartedCreatingCommunity` Msg, which takes that `CreateCommunityData`. To pass in `CreateCommunityData` to `StartedCreatingCommunity`, I'm using the port's `responseData` field, so we need to encode the `CreateCommunityData` when adding the port, and decode it on `jsAddressToMsg` to pass it to `StartedCreatingCommunity`.

I see two approaches we can take on this scenario:
- Do as I described above. This is nice because we don't have the impossible state where the user has submitted the form, and we started creating the community, but the form data isn't valid (e.g. `invitedReward` isn't a valid `Float`). The bad thing is we need to encode and then decode the data (which is why we need `min_balance` when encoding, and why we need this decoder). 
- Use the `Model` data. This is nice because we don't have to encode and then decode `CreateCommunityData`, but we'd have to parse the form data again, which leads to impossible states.

I still prefer passing data through `responseData` in the port, so we could pass in an object like so

```js
{
  createCommunityData : { /* Regular CreateCommunityData, as it was before this PR, without `min_balance` */ },
  minBalance: 0
}
```

That would mean we can keep the encoder without `min_balance`, and avoid those impossible states

Reply:
We do use `newCommunity` subscription, yes, but that's only to get a callback from the server that event-source was able to capture the transaction.

The steps which we take:

1. Grab data from the user, validate form
2. Create a subscription before we actually save anything
3. Push the transaction to the blockchain with two actions (one creates the community and the other the token)
4. Wait for the event-source to capture it
5. Backend has a trigger configured, every time something is inserted to the `community` table it looks for subscriptions and publish them
6. The frontend grabs the event and redirects the user to the newly created community.

Those steps were created to ensure the user would be redirected to the community page after creating them, due to the async nature of the blockchain, if we redirect the user right away sometimes it may happen that they see a 404. This is not ideal but it means the GraphQL subscription is only something related to the automatic redirect, not the saving itself. The point is: to create community you don't need the `min_balance`. You will need it to create a token. 

This is why I find it wierd, that behavior haven't changed even after you added new forms and pages, so no need to have a new decoder to reach the same result we had before

Reply:
The old flow for creating a community was like this:

1. User sees empty form
2. User fills in form
3. User clicks "Save"
4. If the symbol field is actually a valid symbol, we subscribe to GraphQL's `newCommunity` (otherwise, literally do nothing)
5. We wait until the subscription port sends us something back.
6. a. If it sends "responded", just redirect to the community page (and the flow is over)
    b. If it sends "starting", we ask the user for authentication
7. Only now we validate the form
8. a. If the form is not valid, we show the errors to the user (and the flow starts from 2.)
    b. If the form is valid, we send the Eos transaction
9. a. If the Eos response was valid, redirect to community (we should remove this, so we only redirect when the subscription tells us to)
    b. Otherwise, show errors and start from 2.
    
The new flow is like this:

1. User sees empty form
2. User fills in form
3. User clicks "Save"
4. a. If the form is not valid, we show errors (and the flow starts from 2.)
    b. If the form is valid, we ask the user for authentication
5. We subscribe to `newCommunity`
6. We wait until the subscription port sends us something back
7. a. If it sends "responded", redirect to the community (and the flow is over)
    b. If it sends "starting" we send the Eos transaction
8. a. If the Eos response was valid, redirect to community (we should remove this, so we only redirect when the subscription tells us to)
    b. Otherwise, show errors and start from 2.
    
So the old flow validates the model very late (after subscribing to `newCommunity` and asking the user for authentication - IMO it should be before those), while the new one validates it as soon as the user clicks "Save". Because of that, we have the validated information on the `SubmittedForm` msg, not only just before sending the Eos transaction. So we didn't need the decoder before because we only validated the data when sending the Eos transaction - which means it would take longer to show errors to the user.

The result ends up being the same, but the process is a little different, and better for the user. By the way, `minBalance` already was in `CreateCommunityData`, it just wasn't included in the encoder. I'm still convinced `minBalance` should be in `CreateCommunityData` - that's an abstraction we make throughout the app (we don't have a `Community` and a `Token`, the token information is already included in `Community`). Having `min_balance` in the encoder doesn't seem to break anything, but we could make it so the encoder actually returns two encoders (in a tuple), one for the community and one for the token - I think this makes sense, because when we're creating a new community, we're also creating a new token anyway

Reply:
bro, first of all kudos to the complete and amazing explanation. I see this as a good example to be followed and thanks for being this reference!

I've loved the new flow. It is indeed much better. I wonder how we could use Elm to define state machines like this. Anyway...

You've mentioned something that I wasn't aware of: that its possible to have the `responded` reply if the community is already created. This means that we could use other stuff like `transaction_id` in place of `symbol`. This would give us a generic subscription that guarantees transaction pushing and processing to the frontend.  Because now we have multiple subscriptions (creatingCommunity and transfers are the ones I remember). With your confirmation it means it's possible to us to create this. [I've opened an issue on the backend](https://github.com/cambiatus/backend/issues/148), might change the whole responsiveness of the app!

> Having min_balance in the encoder doesn't seem to break anything, but we could make it so the encoder actually returns two encoders (in a tuple), one for the community and one for the token - I think this makes sense, because when we're creating a new community, we're also creating a new token anyway

About this, I think this the right direction, adding the token data to that type. I still prefer to have one type for creating community and another for token, and treat them differently. Messing them both up will lead to an even confusing codebase because the frontend will have a different view of what an community is than on the blockchain. Changes in the contract may hurt a lot in the future if the frontend is not on par with it

Reply:
> I wonder how we could use Elm to define state machines like this.

To me, Elm is pretty much exactly like a state machine (which is a topic I really really like - In fact my first ever Elm program was a [state machine simulator](https://neovier.github.io/LFC01/) - plus some other stuff - for a class project. You can see the source [here](https://github.com/NeoVier/LFC01) if you want. In there, there are some example input files as well), in which you have a `Model` that represents states and `Msg`s that represent transitions. Anyway...

> You've mentioned something that I wasn't aware of: that its possible to have the responded reply if the community is already created. This means that we could use other stuff like transaction_id in place of symbol. This would give us a generic subscription that guarantees transaction pushing and processing to the frontend. Because now we have multiple subscriptions (creatingCommunity and transfers are the ones I remember). With your confirmation it means it's possible to us to create this. I've opened an issue on the backend, might change the whole responsiveness of the app!

That's an amazing accidental finding! :sweat_smile: Tbh, I'm not really sure if that response is even used by the backend right now, but I see how this could be nice, and help with things like [this](https://github.com/cambiatus/frontend/blob/feature/support-subdomains/src/elm/Page/Community/ActionEditor.elm#L1060-L1061) and [this](https://github.com/cambiatus/frontend/blob/feature/support-subdomains/src/elm/Page/Community/ObjectiveEditor.elm#L382-L383).

> About this, I think this the right direction, adding the token data to that type. I still prefer to have one type for creating community and another for token, and treat them differently. Messing them both up will lead to an even confusing codebase because the frontend will have a different view of what an community is than on the blockchain. Changes in the contract may hurt a lot in the future if the frontend is not on par with it

Yes, I guess separating these appropriately can prevent future headaches. As this is a "lower layer" inside the frontend (it's closer to the blockchain than to the user), I think it makes sense to follow the way we're doing it in the blockchain. The `Community.Model` is closer to the user, so there it makes sense to have the community and the token data together - it's an abstraction the user does. In their point of view, the community and the token are kind of the same thing. Let's do it! :rocket:

Reply:
> To me, Elm is pretty much exactly like a state machine (which is a topic I really really like - In fact my first ever Elm program was a state machine simulator - plus some other stuff - for a class project. You can see the source here if you want. In there, there are some example input files as well), in which you have a Model that represents states and Msgs that represent transitions. Anyway...

Nice bro! I think we could improve our Msgs to be more explicit, right now we need to check the `update` function to know what happens next... Just throwing out ideas to have more expressive flows on our pages

>That's an amazing accidental finding! 😅 Tbh, I'm not really sure if that response is even used by the backend right now, but I see how this could be nice, and help with things like this and this.

I've moved [this issue](https://github.com/cambiatus/backend/issues/148) from improvements proposals, we can work on it when possible!